File: core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/promisehandlers/UpdateInputLinkingHandler.scala
package edu.uci.ics.amber.engine.architecture.worker.promisehandlers

import edu.uci.ics.amber.engine.architecture.worker.DataProcessorRPCHandlerInitializer
import edu.uci.ics.amber.engine.architecture.worker.promisehandlers.UpdateInputLinkingHandler.UpdateInputLinking
import edu.uci.ics.amber.engine.architecture.worker.statistics.WorkerState.{PAUSED, READY, RUNNING}
import edu.uci.ics.amber.engine.common.rpc.AsyncRPCServer.ControlCommand
import edu.uci.ics.amber.engine.common.virtualidentity.ActorVirtualIdentity
import edu.uci.ics.amber.engine.common.workflow.PhysicalLink

object UpdateInputLinkingHandler {

  final case class UpdateInputLinking(
      identifier: ActorVirtualIdentity,
      inputLink: PhysicalLink
  ) extends ControlCommand[Unit]
}

trait UpdateInputLinkingHandler {
  this: DataProcessorRPCHandlerInitializer =>

  registerHandler { (msg: UpdateInputLinking, sender) =>
    dp.stateManager.assertState(READY, RUNNING, PAUSED)
    dp.registerInput(msg.identifier, msg.inputLink)
    dp.upstreamLinkStatus.registerInput(msg.identifier, msg.inputLink)
  }

}

File: core/amber/src/main/python/core/architecture/handlers/update_input_linking_handler.py
from core.architecture.handlers.handler_base import Handler
from core.architecture.managers.context import Context
from proto.edu.uci.ics.amber.engine.architecture.worker import UpdateInputLinkingV2


class UpdateInputLinkingHandler(Handler):
    cmd = UpdateInputLinkingV2

    def __call__(
        self, context: Context, command: UpdateInputLinkingV2, *args, **kwargs
    ):
        context.batch_to_tuple_converter.register_input(
            command.identifier, command.input_link
        )
        return None
